### PR TITLE
BUG: Xorbits.numpy raises TypeError 

### LINKED ISSUE #318: BUG: Xorbits.numpy raises TypeError
Reproduce:
```import xorbits
import xorbits
xorbits.init()
from xorbits import numpy as np
a=np.ones(5,dtype=bool)
b=np.zeros(5,dtype=bool)
if np.any(a&b):
    print(1)
```

Full stack:
```
IndexError                                Traceback (most recent call last)
File ~/Documents/projects/xorbits/python/xorbits/core/data.py:221, in DataRef.__len__(self)
    220 try:
--> 221     return int(self.data._mars_entity.shape[0])
    222 except (IndexError, ValueError):
    223     # shape is unknown, execute it

IndexError: tuple index out of range

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
File ~/Documents/projects/xorbits/python/xorbits/core/data.py:226, in DataRef.__len__(self)
    225 try:
--> 226     return int(self.data._mars_entity.shape[0])
    227 except (IndexError, ValueError):  # happens when dimension is 0

IndexError: tuple index out of range

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[2], line 5
      3 a=np.ones(5,dtype=bool)
      4 b=np.zeros(5,dtype=bool)
----> 5 if np.any(a&b):
      6     print(1)

File ~/Documents/projects/xorbits/python/xorbits/core/data.py:228, in DataRef.__len__(self)
    226             return int(self.data._mars_entity.shape[0])
    227         except (IndexError, ValueError):  # happens when dimension is 0
--> 228             raise TypeError(
    229                 f"object with shape {self.data._mars_entity.shape} has no len()"
    230             )
    231 else:
    232     raise TypeError(f"object of type '{self.data.data_type}' has no len()")

TypeError: object with shape () has no len()
```

OS: Windows 10 
python: Python 3.8.8
xorbits: 0.2.1

### LINKED ISSUE #321: ENH: implement magics for auto conversion
### Is your feature request related to a problem? Please describe
Current you need to convert an Xorbits object to python object explicitly.

For example:
```python
df = pd.DataFrame({"foo": [1, 2, 3], "bar": [4, 5, 6]})
max_ref = df.values.max()  # type: DataRef
```

You need to call `to_numpy` explicitly before using `max_ref` as a scalar value:
```python
max_value = max_ref.to_numpy()
```

Errors happen when you use `max_ref` directly:

```python
for i in range(max_ref):
    print(i)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[25], line 1
----> 1 for i in range(max_ref):
      2     print(i)

TypeError: 'DataRef' object cannot be interpreted as an integer
```

### Describe the solution you'd like
For users' convenience, we should let Xorbits handle the type conversion. It can be achieved by implementing magics like `__index__`, `__int__`, `__bool__`.

### PR DESCRIPTION
**Pull request by Zhou1213CN**
## What do these changes do?
Add auto bool and float conversion

## Related issue number

Fixes #318 
Fixes #321 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/core/data.py [hunk 1]
```diff
@@ -40,6 +41,12 @@ class DataType(Enum):
 DATA_MEMBERS: Dict[DataType, Dict[str, Any]] = defaultdict(dict)
 
 
+class ConversionType(Enum):
+    int_conversion = 1
+    float_conversion = 2
+    bool_conversion = 3
+
```
Comment (aresnow1) [inline python/xorbits/core/data.py:48]: Define a method to convert value would be more clean.
``` Python
class ConversionType(Enum):
    int_conversion = 1
    float_conversion = 2
    bool_conversion = 3

    def convert(self, val):
        if self.value == 1:
            return int(val)
        ...
```
For line273 to line 278, just call `conversion.convert(self.to_numpy())`.

File: python/xorbits/core/data.py [hunk 2]
```diff
@@ -255,41 +262,91 @@ def __repr__(self):
             run(self)
             return self.data.__repr__()
 
-    def _to_int(self, cast=False):
+    def _magic_methods(self, conversion, cast=False):
```
Comment (aresnow1) [inline python/xorbits/core/data.py:265]: Rename to `_convert_to`?

**Review by aresnow1 — APPROVED** [review]
LGTM

File: python/xorbits/core/data.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, List, Type
 
 from pandas.api.types import is_integer_dtype
+from pandas.core.dtypes.common import is_float_dtype
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:21]: The `is_float_dtype` can be imported from `pandas.api.types` as well. Normally, the `pandas.core` should not be imported which mentioned in pandas doc that, the core is not for users, api is.

File: python/xorbits/core/data.py [hunk 4]
```diff
@@ -40,6 +41,23 @@ class DataType(Enum):
 DATA_MEMBERS: Dict[DataType, Dict[str, Any]] = defaultdict(dict)
 
 
+class AutoConversionType(Enum):
+    int_conversion = 1
+    float_conversion = 2
+
+    def convert(self, val):
+        if self.value == 1:
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:49]: Better to use `if self == AutoConversionType.int_conversion`.

File: python/xorbits/core/data.py [hunk 5]
```diff
@@ -255,37 +273,66 @@ def __repr__(self):
...
+            if len(self.shape) == 0:
+                run(self)
+                return bool(self.to_numpy())
+            else:
+                if self.__len__() <= 1:
+                    run(self)
+                    return bool(self.to_numpy())
+                else:
+                    raise ValueError(
+                        f"ValueError: The truth value of a {data_type} with more than one element is ambiguous. Use a.any() or a.all()"
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:311]: This line is a bit long, you can split them manually.

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**